### ISSUE TITLE
Bug: no-bare-urls puts closing angle bracket before link's trailing parentheses

### ISSUE THREAD

**Issue by jship** [platers/obsidian-linter#908]
- [X] I have verified that I am on the latest version of the Linter

## Describe the Bug

When using the `no-bare-urls` rule on a link definition like `https://en.wikipedia.org/wiki/The_Score_(album)`, the closing angle bracket will get inserted before the ending parentheses instead of getting inserted at the end.

## How to Reproduce

Put the following into an obsidian document:

```
[The Score]: https://en.wikipedia.org/wiki/The_Score_(album)
```

Then run the linter. The new format in the document will be:

```
[The Score]: <https://en.wikipedia.org/wiki/The_Score_(album>)
```

## Expected Behavior

Upon running the linter, it is expected that the closing angle bracket is put at the end:

``` markdown
[The Score]: <https://en.wikipedia.org/wiki/The_Score_(album)>
```

## Device

- [X] Desktop
- [ ] Mobile

**Comment by zhouxinghong** [comment]
+1

**Comment by jship** [comment]
This behavior also occurs for links with `!` at the end. Linting this link:

```
[This Is Fats Domino!]: https://en.wikipedia.org/wiki/This_Is_Fats_Domino!
```

Results in the closing parentheses being inserted before the `!`:

```
[This Is Fats Domino!]: <https://en.wikipedia.org/wiki/This_Is_Fats_Domino>!
```

**Comment by pjkaufman** [comment]
I will take a look at this and see what is going on. It looks like it is triggering the logic where it thinks it is wrapped in parentheses for the first issue. The issue with ti ending in an exclamation mark may be related but different. I need to make sure that those URLs are valid and properly encoded (my guess is they are already, but I want to make sure before getting too far into this issue.

**Comment by pjkaufman** [comment]
This is being caused by the URL regex since it tries to limit the amount of bad scenarios it encounters. I need to double check the old URL regex against the regex used now and see if it would have worked with the old regex. If not, this would be a new encounter.

One thing to note is that URLs are very complex especially when used in conjunction with punctuation so it may not be possible to perfectly handle punctuation at the end of a URL since it may come down to whether or not the Linter has to filter out punctuation for a sentence versus punctuation in a link.

**Comment by pjkaufman** [comment]
I believe I have a fix for the paren issue. But I want to make sure that the logic makes sense. @jship , here is what I came up with for the logic:
- Check if the link has an opening paren in it
- If there is a different in the count of opening and closing parentheses, then check if the next character after the current URL selection is a closing paren
- If that is the case, include that paren in the URL

This means that 
``` markdown
[The Score]: https://en.wikipedia.org/wiki/The_Score_(album)
```
becomes 
``` markdown
[The Score]: <https://en.wikipedia.org/wiki/The_Score_(album)>
```
Where as the following
``` markdown
[The Score]: https://en.wikipedia.org/wiki/The_Score_album)
```
becomes 
``` markdown
[The Score]: <https://en.wikipedia.org/wiki/The_Score_album>)
```

Does this look correct? If not, do you have any suggestions about how to tell if the closing paren is supposed to be part of the URL (http requests are not a really a valid option for this)?

**Comment by pjkaufman** [comment]
Also, how are you suggesting that the Linter differentiate regular punctuation from URL punctuation?

`!` is not URL encoded so it gets considered to be sentence punctuation. That can be changed, but that means usage in sentences will likely get messed up.

**Comment by pjkaufman** [comment]
Well, it looks like the old regex handled the punctuation properly, so I may just need to see why it is not currently being handled correctly instead of trying to add extra logic into the scenario.

**Comment by pjkaufman** [comment]
I believe I have a fix. It just fixes up the regex a little. However it will not allow URLs to end in periods. It will however allow some other punctuation.

**Comment by jship** [comment]
Thank you! I believe the fix you described of handling some punctuation characters but not periods sounds very reasonable and will likely cover the majority of URLs.

**Comment by pjkaufman** [comment]
I plan to push out a bug fix for something soon, so this change should be included in that.

**Comment by pjkaufman** [comment]
This should be on master and in the next release. Please let us know if there is an issue on master or in the next release.
